Move LambdaRewriter MakeFrames into Analysis

agocke opened the pull request:
This PR is dependent on https://github.com/dotnet/roslyn/pull/21088.

This is part of the effort to move as much logic from LambdaRewriter
into Analysis as possible. The more logic there is in the Rewriter, the
more difficult it is to find calculation problems until the last
possible moment, and it's also very difficult to calculate useful
information for later analysis passes, like the final debug and closure
IDs for each environment.

cston commented on src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs:
```diff
@@ -106,6 +114,9 @@ public sealed class Closure
                 public PooledHashSet<Symbol> CapturedVariables { get; } = PooledHashSet<Symbol>.GetInstance();
 
+                public ArrayBuilder<ClosureEnvironment> CapturedEnvironments { get; internal set; }
+                    = ArrayBuilder<ClosureEnvironment>.GetInstance();
```
Is setter needed?

cston commented on src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs:
```diff
@@ -115,6 +126,38 @@ public Closure(MethodSymbol symbol)
+            /// Represents an environment that holds all captured variables
+            /// from nested functions that are lifted into fields.
+            /// </summary>
+            public sealed class ClosureEnvironment
```
`private`?

cston commented on src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs:
```diff
@@ -115,6 +126,38 @@ public Closure(MethodSymbol symbol)
+
+                public DebugId MethodId { get; } 
+
+                public DebugId ClosureId { get; }
```
Why readonly properties rather than fields?

cston commented on src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs:
```diff
@@ -337,158 +345,51 @@ protected override NamedTypeSymbol ContainingType
+        /// to the compilation.
         /// </summary>
-        private void MakeFrames(ArrayBuilder<ClosureDebugInfo> closureDebugInfo)
+        internal void MakeFramesFromEnvironments()
```
`private`?

cston commented on src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.cs:
```diff
@@ -231,6 +268,83 @@ void RecordClosureScope(Scope innermost, Scope outermost, Closure closure)
+                    var canBeStruct = !ScopesThatCantBeStructs.Contains(scopeBoundNode);
+
+                    var containingMethod = scope.ContainingClosureOpt?.OriginalMethodSymbol ?? _topLevelMethod;
+                    if (_substitutedSourceMethod != null && containingMethod == _topLevelMethod)
```
`(object)_substitutedSourceMethod != null`

cston commented on src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.cs:
```diff
@@ -337,158 +345,51 @@ protected override NamedTypeSymbol ContainingType
+                        env.ClosureId);
+
+                    CompilationState.ModuleBuilderOpt.AddSynthesizedDefinition(ContainingType, frame);
+                    if (frame.Constructor != null)
```
`(object)frame.Constructor != null`

cston commented:
LGTM, minor comments only.

agocke commented:
@cston I've addressed your comments, but unfortunately as I was making more progress I decided I had kind of gone down the wrong path in this PR, so I've gone back and fixed it up with some of the decisions I made later. Most notably, instead of creating a new ClosureEnvironment type I've just repurposed the LambdaFrame type as `ClosureEnvironment`.

jaredpar commented on src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaFrame.cs:
```diff
@@ -1,44 +1,58 @@
+        /// Useful because this frame's type parameters are constructed from this method and all methods containing this method.
+        /// </summary>
+        internal readonly MethodSymbol ContainingMethod;
+        internal readonly FieldSymbol SingletonCache;
```
This type uses a mix of readonly fields and properties. We should be consistent here.

khyperia commented on src/Compilers/Core/Portable/InternalUtilities/ImmutableArrayExtensions.cs:
```diff
@@ -9,24 +9,7 @@ namespace Roslyn.Utilities
-
-            return items;
-        }
+            => items == null ? ImmutableArray<T>.Empty : ImmutableArray.CreateRange(items);
```
I would suggest this goes into master for perf (in case there wasn't an overload for zero-args and we were allocating a params array), but turns out that [`ImmutableArray.Create()`](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray.cs#L30) calls through to `ImmutableArray.Empty`, so this is fine in this branch.

Also to note (to other reviewers), `ImmutableArray.CreateRange` does a dynamic check to see if the parameter [is a boxed ImmutableArray](https://github.com/dotnet/corefx/blob/master/src/System.Collections.Immutable/src/System/Collections/Immutable/ImmutableArray.cs#L102), so there shouldn't be any perf issues from overloads suddenly calling `CreateRange` instead of passing back the same array in the other overload.

jaredpar commented on src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaFrame.cs:
```diff
@@ -1,44 +1,58 @@
+        /// The closest method/lambda that this frame is originally from. Null if nongeneric static closure.
+        /// Useful because this frame's type parameters are constructed from this method and all methods containing this method.
+        /// </summary>
+        internal readonly MethodSymbol ContainingMethod;
```
`OriginalContainingMethod`?

jaredpar commented on src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaRewriter.Analysis.Tree.cs:
```diff
@@ -25,14 +26,14 @@ internal sealed partial class Analysis : BoundTreeWalkerWithStackGuardWithoutRec
             public sealed class Scope
             {
-                public Scope Parent { get; }
+                public readonly Scope Parent;
```
Why this set of changes?

khyperia commented on src/Compilers/Core/Portable/InternalUtilities/SetWithInsertionOrder.cs:
```diff
@@ -16,46 +17,45 @@ namespace Roslyn.Utilities
-        private uint _nextElementValue = 0;
-        private T[] _elements = null;
+        private HashSet<T> _set = null;
+        private ArrayBuilder<T> _elements = null;
```
Yay not reinventing stuff! 🎉

agocke commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs:
```diff
@@ -98,15 +120,14 @@ .maxstack  2
 }");
             // Skip some... L5
-            verifier.VerifyIL("C.<M>g__L52_4(ref C.<>c__DisplayClass2_0, ref C.<>c__DisplayClass2_1)", @"
+            verifier.VerifyIL("C.<M>g__L32_2(ref C.<>c__DisplayClass2_0)", @"
```
Woops, I meant to include L5 here. I'll re-add.

ivanbasov commented on src/Compilers/CSharp/Portable/Lowering/LambdaRewriter/LambdaFrame.cs:
```diff
@@ -1,44 +1,58 @@
+    /// captured by nested functions within that scope.
     /// </summary>
-    internal sealed class LambdaFrame : SynthesizedContainer, ISynthesizedMethodBodyImplementationSymbol
+    internal sealed class ClosureEnvironment : SynthesizedContainer, ISynthesizedMethodBodyImplementationSymbol
```
>ClosureEnvironment [](start = 26, length = 18)

It seems the file should be renamed as well.

agocke replied:
Yeah there are a bunch of name changes. I'll get to it later.

agocke commented:
@khyperia I think I've resolved all your comments -- anything more?

khyperia commented on src/Compilers/Core/Portable/InternalUtilities/ImmutableArrayExtensions.cs:
```diff
@@ -9,24 +9,10 @@ namespace Roslyn.Utilities
-
-            return items;
-        }
+            => items.IsDefault ? ImmutableArray<T>.Empty : ImmutableArray.CreateRange(items);
```
I think this still boxes `items` to `IEnumerable<T>`, I don't see an overload `ImmutableArray.CreateRange(ImmutableArray<T>)`. Probably want to just return the array as-is: `items.IsDefault ? ImmutableArray<T>.Empty : items`

agocke replied:
Hah that's just dumb, will fix.
